File: ios/Plugin/FacebookSharePlugin.swift
import Foundation
import Capacitor
import FBSDKShareKit
/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(FacebookSharePlugin)
public class FacebookSharePlugin: CAPPlugin {
    private var implementation = FacebookShare()
    private var callbackId = ""

    override public func load() {
        implementation.plugin = self
    }

    @objc func sharePhoto(_ call: CAPPluginCall) {
        let data = call.getString("data") ?? ""
        let hashtags = call.getString("hashtags") ?? ""
        let sharedMode = call.getString("sharedMode", "automatic")
        implementation.sharePhoto(data, withHashtags: hashtags, withSharedMode: sharedMode)
        callbackId = call.callbackId
        bridge?.saveCall(call)
    }

    @objc func shareVideo(_ call: CAPPluginCall) {
        let path = call.getString("path") ?? ""
        let hashtags = call.getString("hashtags") ?? ""
        let sharedMode = call.getString("sharedMode", "automatic")
        implementation.shareVideo(path, withHashtags: hashtags, withSharedMode: sharedMode)
        callbackId = call.callbackId
        bridge?.saveCall(call)
    }
}

extension FacebookSharePlugin: SharingDelegate {

    public func sharer(_ sharer: Sharing, didCompleteWithResults results: [String: Any]) {
        var response = JSObject()
        response["success"] = true
        response["message"] = "Shared successfully"

        proccessResponse(response)

    }

    public func sharer(_ sharer: Sharing, didFailWithError error: Error) {
        var response = JSObject()
        response["success"] = false
        response["message"] = error.localizedDescription

        proccessResponse(response)
    }

    public func sharerDidCancel(_ sharer: Sharing) {
        var response = JSObject()
        response["success"] = false
        response["message"] = "User cancelled"

        proccessResponse(response)
    }

    private func proccessResponse(_ response: JSObject) {
        if let savedCall = bridge?.savedCall(withID: callbackId) {
            savedCall.resolve(response)
            bridge?.releaseCall(savedCall)
        }
    }
}

File: android/src/main/java/com/apestudios/facebook/share/FacebookSharePlugin.java
package com.apestudios.facebook.share;

import android.content.Intent;
import android.util.Log;
import androidx.annotation.NonNull;
import com.facebook.CallbackManager;
import com.facebook.FacebookCallback;
import com.facebook.FacebookException;
import com.facebook.FacebookSdk;
import com.facebook.share.Sharer;
import com.facebook.share.widget.ShareDialog;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

@CapacitorPlugin(name = "FacebookShare")
public class FacebookSharePlugin extends Plugin {

    private FacebookShare implementation;

    private String callbackId;

    private CallbackManager callbackManager;
    private ShareDialog shareDialog;

    private FacebookCallback<Sharer.Result> shareCallback = new FacebookCallback<Sharer.Result>() {
        @Override
        public void onCancel() {
            Log.d("FacebookSharePlugin", "Canceled");

            PluginCall savedCall = getBridge().getSavedCall(callbackId);

            if (savedCall == null) {
                Log.e(getLogTag(), "error: no plugin saved call found.");
                return;
            }

            JSObject response = new JSObject();
            response.put("success", false);
            response.put("message", "User cancelled");
            savedCall.resolve(response);
            getBridge().releaseCall(callbackId);
        }

        @Override
        public void onError(@NonNull FacebookException error) {
            Log.d(getLogTag(), String.format("Error: %s", error.toString()));

            PluginCall savedCall = getBridge().getSavedCall(callbackId);

            if (savedCall == null) {
                Log.e(getLogTag(), "error: no plugin saved call found.");
                return;
            }

            JSObject response = new JSObject();
            response.put("success", false);
            response.put("message", error.getMessage());
            savedCall.resolve(response);
            getBridge().releaseCall(callbackId);
        }

        @Override
        public void onSuccess(@NonNull Sharer.Result result) {
            Log.d("FacebookSharePlugin", "Success!");
            Log.w(getLogTag(), "callbackIdSaved: " + callbackId);
            PluginCall savedCall = getBridge().getSavedCall(callbackId);

            if (savedCall == null) {
                Log.e(getLogTag(), "error: no plugin saved call found.");
                return;
            }

            JSObject response = new JSObject();
            response.put("success", true);
            response.put("message", "Shared successfully");
            if (result.getPostId() != null) {
                response.put("postId", result.getPostId());
            }
            savedCall.resolve(response);
            getBridge().releaseCall(callbackId);
        }
    };

    @Override
    public void load() {
        FacebookSdk.sdkInitialize(getActivity().getApplicationContext());
        callbackManager = CallbackManager.Factory.create();
        shareDialog = new ShareDialog(getActivity());
        shareDialog.registerCallback(callbackManager, shareCallback);
        implementation = new FacebookShare(shareDialog);
    }

    @PluginMethod
    public void sharePhoto(PluginCall call) {
        String data = call.getString("data");
        String hashtags = call.getString("hashtags");

        implementation.sharePhoto(data, hashtags);

        callbackId = call.getCallbackId();
        Log.w(getLogTag(), "callbackId: " + callbackId);
        this.bridge.saveCall(call);
    }

    @Override
    protected void handleOnActivityResult(int requestCode, int resultCode, Intent data) {
        Log.d(getLogTag(), "Entering handleOnActivityResult(" + requestCode + ", " + resultCode + ")");

        if (callbackManager.onActivityResult(requestCode, resultCode, data)) {
            Log.d(getLogTag(), "onActivityResult succeeded");
        } else {
            Log.w(getLogTag(), "onActivityResult failed");
        }
    }
}
